File: sdks/dart-server/lib/src/databases.dart
import 'client.dart';

/// Databases service — manage databases, tables, columns, indexes, and rows.
class Databases {
  final ApplAdServer _client;

  Databases(this._client);

  // --- Databases ---

  /// Create a new database.
  Future<Map<String, dynamic>> createDatabase({
    required String name,
    String? databaseId,
  }) async {
    return _client.post('/v1/databases', data: {
      'name': name,
      'databaseId': databaseId ?? 'unique()',
    });
  }

  /// List all databases.
  Future<Map<String, dynamic>> listDatabases() async {
    return _client.get('/v1/databases');
  }

  /// Get a database by ID.
  Future<Map<String, dynamic>> getDatabase(String databaseId) async {
    return _client.get('/v1/databases/$databaseId');
  }

  /// Update a database.
  Future<Map<String, dynamic>> updateDatabase(
      String databaseId, String name) async {
    return _client.put('/v1/databases/$databaseId', data: {
      'name': name,
    });
  }

  /// Delete a database.
  Future<Map<String, dynamic>> deleteDatabase(String databaseId) async {
    return _client.delete('/v1/databases/$databaseId');
  }

  // --- Tables ---

  /// Create a table.
  Future<Map<String, dynamic>> createTable({
    required String databaseId,
    required String name,
    String? tableId,
    List<String>? permissions,
    bool documentSecurity = false,
  }) async {
    return _client.post('/v1/databases/$databaseId/tables', data: {
      'name': name,
      'tableId': tableId ?? 'unique()',
      'permissions': permissions ?? [],
      'documentSecurity': documentSecurity,
    });
  }

  /// List tables in a database.
  Future<Map<String, dynamic>> listTables(String databaseId) async {
    return _client.get('/v1/databases/$databaseId/tables');
  }

  /// Get a table.
  Future<Map<String, dynamic>> getTable(
      String databaseId, String tableId) async {
    return _client.get('/v1/databases/$databaseId/tables/$tableId');
  }

  /// Update a table.
  Future<Map<String, dynamic>> updateTable({
    required String databaseId,
    required String tableId,
    required String name,
    List<String>? permissions,
    bool? enabled,
  }) async {
    return _client.put(
      '/v1/databases/$databaseId/tables/$tableId',
      data: {
        'name': name,
        if (permissions != null) 'permissions': permissions,
        if (enabled != null) 'enabled': enabled,
      },
    );
  }

  /// Delete a table.
  Future<Map<String, dynamic>> deleteTable(
      String databaseId, String tableId) async {
    return _client
        .delete('/v1/databases/$databaseId/tables/$tableId');
  }

  // --- Columns ---

  /// Create a string column.
  Future<Map<String, dynamic>> createStringColumn({
    required String databaseId,
    required String tableId,
    required String key,
    bool required_ = false,
    int? size,
    String? defaultValue,
    bool array = false,
  }) async {
    return _client.post(
      '/v1/databases/$databaseId/tables/$tableId/columns/string',
      data: {
        'key': key,
        'required': required_,
        'array': array,
        if (size != null) 'size': size,
        if (defaultValue != null) 'default': defaultValue,
      },
    );
  }

  /// Create an integer column.
  Future<Map<String, dynamic>> createIntegerColumn({
    required String databaseId,
    required String tableId,
    required String key,
    bool required_ = false,
    num? min,
    num? max,
    int? defaultValue,
    bool array = false,
  }) async {
    return _client.post(
      '/v1/databases/$databaseId/tables/$tableId/columns/integer',
      data: {
        'key': key,
        'required': required_,
        'array': array,
        if (min != null) 'min': min,
        if (max != null) 'max': max,
        if (defaultValue != null) 'default': defaultValue,
      },
    );
  }

  /// Create a boolean column.
  Future<Map<String, dynamic>> createBooleanColumn({
    required String databaseId,
    required String tableId,
    required String key,
    bool required_ = false,
    bool? defaultValue,
    bool array = false,
  }) async {
    return _client.post(
      '/v1/databases/$databaseId/tables/$tableId/columns/boolean',
      data: {
        'key': key,
        'required': required_,
        'array': array,
        if (defaultValue != null) 'default': defaultValue,
      },
    );
  }

  /// Create an enum column.
  Future<Map<String, dynamic>> createEnumColumn({
    required String databaseId,
    required String tableId,
    required String key,
    required List<String> elements,
    bool required_ = false,
    String? defaultValue,
    bool array = false,
  }) async {
    return _client.post(
      '/v1/databases/$databaseId/tables/$tableId/columns/enum',
      data: {
        'key': key,
        'required': required_,
        'array': array,
        'elements': elements,
        if (defaultValue != null) 'default': defaultValue,
      },
    );
  }

  /// List columns of a table.
  Future<Map<String, dynamic>> listColumns(
      String databaseId, String tableId) async {
    return _client.get(
        '/v1/databases/$databaseId/tables/$tableId/columns');
  }

  /// Delete a column.
  Future<Map<String, dynamic>> deleteColumn(
      String databaseId, String tableId, String key) async {
    return _client.delete(
        '/v1/databases/$databaseId/tables/$tableId/columns/$key');
  }

  // --- Indexes ---

  /// Create an index.
  Future<Map<String, dynamic>> createIndex({
    required String databaseId,
    required String tableId,
    required String key,
    required String type,
    required List<String> columns,
    List<String>? orders,
  }) async {
    return _client.post(
      '/v1/databases/$databaseId/tables/$tableId/indexes',
      data: {
        'key': key,
        'type': type,
        'columns': columns,
        if (orders != null) 'orders': orders,
      },
    );
  }

  /// List indexes of a table.
  Future<Map<String, dynamic>> listIndexes(
      String databaseId, String tableId) async {
    return _client.get(
        '/v1/databases/$databaseId/tables/$tableId/indexes');
  }

  /// Delete an index.
  Future<Map<String, dynamic>> deleteIndex(
      String databaseId, String tableId, String key) async {
    return _client.delete(
        '/v1/databases/$databaseId/tables/$tableId/indexes/$key');
  }

  // --- Rows ---

  /// Create a row.
  Future<Map<String, dynamic>> createRow({
    required String databaseId,
    required String tableId,
    required Map<String, dynamic> data,
    String? rowId,
    List<String>? permissions,
  }) async {
    return _client.post(
      '/v1/databases/$databaseId/tables/$tableId/rows',
      data: {
        'rowId': rowId ?? 'unique()',
        'data': data,
        'permissions': permissions ?? [],
      },
    );
  }

  /// List rows in a table.
  Future<Map<String, dynamic>> listRows({
    required String databaseId,
    required String tableId,
    int? limit,
    int? offset,
  }) async {
    final query = <String, String>{};
    if (limit != null) query['limit'] = limit.toString();
    if (offset != null) query['offset'] = offset.toString();
    final qs = query.isNotEmpty
        ? '?${query.entries.map((e) => '${e.key}=${Uri.encodeComponent(e.value)}').join('&')}'
        : '';
    return _client.get(
        '/v1/databases/$databaseId/tables/$tableId/rows$qs');
  }

  /// Get a row by ID.
  Future<Map<String, dynamic>> getRow({
    required String databaseId,
    required String tableId,
    required String rowId,
  }) async {
    return _client.get(
        '/v1/databases/$databaseId/tables/$tableId/rows/$rowId');
  }

  /// Update a row.
  Future<Map<String, dynamic>> updateRow({
    required String databaseId,
    required String tableId,
    required String rowId,
    Map<String, dynamic>? data,
    List<String>? permissions,
  }) async {
    return _client.patch(
      '/v1/databases/$databaseId/tables/$tableId/rows/$rowId',
      data: {
        if (data != null) 'data': data,
        if (permissions != null) 'permissions': permissions,
      },
    );
  }

  /// Delete a row.
  Future<Map<String, dynamic>> deleteRow({
    required String databaseId,
    required String tableId,
    required String rowId,
  }) async {
    return _client.delete(
        '/v1/databases/$databaseId/tables/$tableId/rows/$rowId');
  }
}

File: sdks/react-native/src/databases.ts
import type { ApplAdClient } from './client';

export class Databases {
  constructor(private client: ApplAdClient) {}

  // --- Databases ---

  createDatabase(name: string, databaseId?: string) {
    return this.client.call('POST', '/databases', {
      name,
      databaseId: databaseId ?? 'unique()',
    });
  }

  listDatabases() {
    return this.client.call('GET', '/databases');
  }

  getDatabase(databaseId: string) {
    return this.client.call('GET', `/databases/${databaseId}`);
  }

  updateDatabase(databaseId: string, name: string) {
    return this.client.call('PUT', `/databases/${databaseId}`, { name });
  }

  deleteDatabase(databaseId: string) {
    return this.client.call('DELETE', `/databases/${databaseId}`);
  }

  // --- Tables ---

  createTable(
    databaseId: string,
    name: string,
    opts?: { tableId?: string; permissions?: string[]; documentSecurity?: boolean }
  ) {
    return this.client.call('POST', `/databases/${databaseId}/tables`, {
      name,
      tableId: opts?.tableId ?? 'unique()',
      permissions: opts?.permissions ?? [],
      documentSecurity: opts?.documentSecurity ?? false,
    });
  }

  listTables(databaseId: string) {
    return this.client.call('GET', `/databases/${databaseId}/tables`);
  }

  getTable(databaseId: string, tableId: string) {
    return this.client.call('GET', `/databases/${databaseId}/tables/${tableId}`);
  }

  updateTable(
    databaseId: string,
    tableId: string,
    name: string,
    opts?: { permissions?: string[]; enabled?: boolean }
  ) {
    return this.client.call('PUT', `/databases/${databaseId}/tables/${tableId}`, {
      name,
      ...(opts?.permissions && { permissions: opts.permissions }),
      ...(opts?.enabled !== undefined && { enabled: opts.enabled }),
    });
  }

  deleteTable(databaseId: string, tableId: string) {
    return this.client.call('DELETE', `/databases/${databaseId}/tables/${tableId}`);
  }

  // --- Columns ---

  createStringColumn(
    databaseId: string,
    tableId: string,
    key: string,
    opts?: { required?: boolean; size?: number; defaultValue?: string; array?: boolean }
  ) {
    return this.client.call(
      'POST',
      `/databases/${databaseId}/tables/${tableId}/columns/string`,
      {
        key,
        required: opts?.required ?? false,
        array: opts?.array ?? false,
        ...(opts?.size && { size: opts.size }),
        ...(opts?.defaultValue !== undefined && { default: opts.defaultValue }),
      }
    );
  }

  createIntegerColumn(
    databaseId: string,
    tableId: string,
    key: string,
    opts?: { required?: boolean; min?: number; max?: number; defaultValue?: number; array?: boolean }
  ) {
    return this.client.call(
      'POST',
      `/databases/${databaseId}/tables/${tableId}/columns/integer`,
      {
        key,
        required: opts?.required ?? false,
        array: opts?.array ?? false,
        ...(opts?.min !== undefined && { min: opts.min }),
        ...(opts?.max !== undefined && { max: opts.max }),
        ...(opts?.defaultValue !== undefined && { default: opts.defaultValue }),
      }
    );
  }

  createBooleanColumn(
    databaseId: string,
    tableId: string,
    key: string,
    opts?: { required?: boolean; defaultValue?: boolean; array?: boolean }
  ) {
    return this.client.call(
      'POST',
      `/databases/${databaseId}/tables/${tableId}/columns/boolean`,
      {
        key,
        required: opts?.required ?? false,
        array: opts?.array ?? false,
        ...(opts?.defaultValue !== undefined && { default: opts.defaultValue }),
      }
    );
  }

  listColumns(databaseId: string, tableId: string) {
    return this.client.call('GET', `/databases/${databaseId}/tables/${tableId}/columns`);
  }

  deleteColumn(databaseId: string, tableId: string, key: string) {
    return this.client.call('DELETE', `/databases/${databaseId}/tables/${tableId}/columns/${key}`);
  }

  // --- Indexes ---

  createIndex(
    databaseId: string,
    tableId: string,
    key: string,
    type: string,
    columns: string[],
    orders?: string[]
  ) {
    return this.client.call(
      'POST',
      `/databases/${databaseId}/tables/${tableId}/indexes`,
      { key, type, columns, ...(orders && { orders }) }
    );
  }

  listIndexes(databaseId: string, tableId: string) {
    return this.client.call('GET', `/databases/${databaseId}/tables/${tableId}/indexes`);
  }

  deleteIndex(databaseId: string, tableId: string, key: string) {
    return this.client.call('DELETE', `/databases/${databaseId}/tables/${tableId}/indexes/${key}`);
  }

  // --- Rows (Documents) ---

  createDocument(
    databaseId: string,
    tableId: string,
    data: Record<string, unknown>,
    opts?: { rowId?: string; permissions?: string[] }
  ) {
    return this.client.call('POST', `/databases/${databaseId}/tables/${tableId}/rows`, {
      rowId: opts?.rowId ?? 'unique()',
      data,
      permissions: opts?.permissions ?? [],
    });
  }

  listDocuments(
    databaseId: string,
    tableId: string,
    opts?: { limit?: number; offset?: number; queries?: string[] }
  ) {
    const params = new URLSearchParams();
    if (opts?.limit) params.set('limit', String(opts.limit));
    if (opts?.offset) params.set('offset', String(opts.offset));
    if (opts?.queries) {
      opts.queries.forEach((q) => params.append('queries[]', q));
    }
    const qs = params.toString();
    return this.client.call(
      'GET',
      `/databases/${databaseId}/tables/${tableId}/rows${qs ? `?${qs}` : ''}`
    );
  }

  getDocument(databaseId: string, tableId: string, rowId: string) {
    return this.client.call('GET', `/databases/${databaseId}/tables/${tableId}/rows/${rowId}`);
  }

  updateDocument(
    databaseId: string,
    tableId: string,
    rowId: string,
    opts?: { data?: Record<string, unknown>; permissions?: string[] }
  ) {
    return this.client.call(
      'PATCH',
      `/databases/${databaseId}/tables/${tableId}/rows/${rowId}`,
      {
        ...(opts?.data && { data: opts.data }),
        ...(opts?.permissions && { permissions: opts.permissions }),
      }
    );
  }

  deleteDocument(databaseId: string, tableId: string, rowId: string) {
    return this.client.call('DELETE', `/databases/${databaseId}/tables/${tableId}/rows/${rowId}`);
  }

  /** Create or update a document by ID. */
  upsertDocument(
    databaseId: string,
    tableId: string,
    rowId: string,
    data: Record<string, unknown>,
    opts?: { permissions?: string[] }
  ) {
    return this.client.call('PUT', `/databases/${databaseId}/tables/${tableId}/rows/${rowId}`, {
      data,
      permissions: opts?.permissions ?? [],
    });
  }
}
